OTF encoder

camertron opened the pull request:
_**This pull request is part of a larger effort to bring OTF support to TTFunk. See https://github.com/prawnpdf/ttfunk/issues/53 for details.**_

This PR adds the `OTFEncoder` class. There's really no difference between .otf and .ttf fonts except that CFF-based fonts (often given the .otf extension) have a different optimal table order.I

Things to note in this PR:
1. I went through and changed string encoding in a couple of places to make sure there is uniform encoding throughout the CFF code. Since we're packing binary data, the strings emitted should always have ASCII-8BIT encoding.
2. Supplemental CFF encoding support has been added.
3. Added support for the gasp table.
4. Fixed a number of issues I found after parsing around 10,000 real fonts.

@pointlessone unfortunately this won't be the last PR. I have a couple more in the works that I will add to #53. I also just re-tested subsetting Noto, and while all the ASCII characters render in Photoshop, Keynote, etc, none of the Japanese characters work (I just see blank spaces). So unfortunately something's wrong. I will need to identify and fix the issue before we can consider the OTF work done.

pointlessone commented on lib/ttfunk/table/cff.rb:
```diff
@@ -31,18 +31,18 @@ def tag
       end
 
-      def encode(mapping)
+      def encode(new2_old, old2_new)
```
I don't quite get this change. Both mapping appear to always be derivable from another. Why do we need to pass around both?

camertron replied:
Yes they are derivable from each other, but passing them around is much more efficient than requiring each `encode` method to re-create them.

pointlessone replied:
I see. Please document this somewhere. Either a comment in source or commit message.

camertron replied:
Done

camertron commented on lib/ttfunk/table/cff/encoding.rb:
```diff
@@ -89,8 +92,30 @@ def encode(mapping)
+
         private
 
+        def encode_supplemental(_new2_old, old2_new)
```
I think I named these this way because of existing code, should we clean it up everywhere?

pointlessone replied:
Since we're changing this code substantially we might as well clean it up a bit.

camertron replied:
Done (renamed to `old_to_new` and `new_to_old`).

camertron commented on lib/ttfunk/table/maxp.rb:
```diff
@@ -50,6 +52,10 @@ def parse!
             @max_stack_elements, @max_size_of_instructions,
             @max_component_elements, @max_component_depth = read(28, 'Nn*')
+
+          # a number of fonts omit these last two bytes for some reason,
```
I believe so, yes. I saw the `nil` issue in fonts where the maxp table was the last table in the font. We do have the length here so it would not be too difficult to protect against reading past the end of the table.

pointlessone replied:
Let's do that.

camertron replied:
Done (and found a bug as a result!)

camertron commented on lib/ttfunk/table/cff/dict.rb:
```diff
@@ -177,8 +177,8 @@ def decode_operand(b_zero)
         end
 
         def decode_sci
-          significand = []
```
In this case I want a binary string, but I also want it to be mutable. Not sure I understand what your concern is here... mutating strings is not inherently bad is it?

pointlessone replied:
It is not inherently bad to mutate a string. However, I'm trying to avoid it in the codebase for consistency.

camertron commented on lib/ttfunk/encoded_string.rb:
```diff
@@ -51,15 +51,15 @@ def string
       end
 
-      io.string
+      io.string.force_encoding(Encoding::ASCII_8BIT)
```
Ok that makes sense. As I recall `StringIO` doesn't respect the `b` option if you pass it in as the `mode`, which also could have been a solution here.

camertron commented:
Ok @pointlessone I've addressed all your comments and fixed subsetting too. Code can subset Noto and resulting font can be used by word processors, etc.
